libexec/rtld-elf/rtld_c18n.c

```diff
@@ -0,0 +1,700 @@
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+__FBSDID("$FreeBSD$");
+
+#include <sys/param.h>
+#include <sys/mman.h>
+
+#include <stdlib.h>
+#include <stdatomic.h>
+
+#include "debug.h"
+#include "rtld.h"
+#include "rtld_c18n.h"
+#include "rtld_libc.h"
+#include "rtld_utrace.h"
+
+/*
+ * Sealers for RTLD privileged information
+ */
+uintptr_t sealer_pltgot, sealer_jmpbuf, sealer_tramp;
+
+const char *ld_compartment_utrace;	/* Use utrace() to log compartmentalisation-related events */
+const char *ld_compartment_enable;	/* Enable compartmentalisation */
+const char *ld_compartment_overhead;	/* Simulate overhead during compartment transitions */
+
+/*
+ * libthr support
+ */
+void (*thr_thread_start)(struct pthread *);
+
+void
+_rtld_thread_start_init(void (*p)(struct pthread *))
+{
+	assert((cheri_getperm(p) & CHERI_PERM_EXECUTIVE) == 0);
+	assert(thr_thread_start == NULL);
+	thr_thread_start = tramp_intern(NULL, &(struct tramp_data) {
+		.target = p,
+		.defobj = obj_from_addr(p),
+		.sig = (struct tramp_sig) {
+			.valid = true,
+			.reg_args = 1, .mem_args = false, .ret_args = NONE
+		}
+	});
+}
+
+void (*thr_sighandler)(int, siginfo_t *, void *);
+
+void
+_rtld_sighandler_init(void (*p)(int, siginfo_t *, void *))
+{
+	assert((cheri_getperm(p) & CHERI_PERM_EXECUTIVE) == 0);
+	assert(thr_sighandler == NULL);
+	thr_sighandler = tramp_intern(NULL, &(struct tramp_data) {
+		.target = p,
+		.defobj = obj_from_addr(p),
+		.sig = (struct tramp_sig) {
+			.valid = true,
+			.reg_args = 3, .mem_args = false, .ret_args = NONE
+		}
+	});
+}
+
+/*
+ * Policies
+ */
+typedef ssize_t string_handle;
+
+struct string_base {
+	char *buf;
+	string_handle size, capacity;
+};
+
+__unused static struct string_base *
+string_base_create(string_handle capacity)
+{
+	struct string_base *sb = xmalloc(sizeof(*sb));
+	sb->buf = xmalloc(sizeof(*sb->buf) * capacity);
+	sb->size = 0;
+	sb->capacity = capacity;
+	return (sb);
+}
+
+__unused static string_handle
+string_base_push(struct string_base *sb, const char *str)
+{
+	string_handle handle = sb->size;
+	do {
+		sb->buf[sb->size++] = *str;
+		if (sb->size == sb->capacity) {
+			sb->capacity *= 2;
+			sb->buf = realloc(sb->buf,
+			    sizeof(*sb->buf) * sb->capacity);
+			if (sb->buf == NULL)
+				rtld_fatal("realloc failed");
+		}
+	} while (*str++ != '\0');
+	return (handle);
+}
+
+static string_handle
+string_base_search(const struct string_base *sb, const char *str)
+{
+	string_handle i = 0;
+	do {
+		const char *cur = str;
+		while (sb->buf[i] == *cur++)
+			if (sb->buf[i++] == '\0')
+				return (i);
+		while (sb->buf[i] != '\0') ++i;
+	} while (++i < sb->size);
+	return (-1);
+}
+
+static char trusted_globals_names[] =
+	"memset\0"
```

Comment:
What's wrong with a 2D array, or 1D array of pointers

Reply:
They seem too heavyweight for a bunch of small strings.

Reply:
Heavyweight? How? It's cleaner.

Reply:
Well, a pointer is already larger than most of these strings... And I will also need to store up to ~1000 small strings dynamically in a future feature.

Reply:
So do a 2D array then.

Reply:
Hm, I guess that requires the size of each element to be fixed.